### PR TITLE
LX: lx_vdso.so PT_DYNAMIC section should not be writeable 

### LINKED ISSUE #1142: centos-stream 9 / fedora 35 do not boot under lx
Neither of these recent Linux distributions boot under lx, and both behave in the same way. All processes crash in the runtime linker with:

```
status: process terminated by SIGSEGV (Segmentation Fault), addr=7fffef45c100

00007fffffeffc50 ld-linux-x86-64.so.2`dl_main+0x24c7()
00007fffef28a040 ld-linux-x86-64.so.2`_dl_sysdep_start+0x3fc()
0000000400000006 0x40()

> <rip::dis -n0
ld-linux-x86-64.so.2`dl_main+0x24c7:    addq   %rcx,0x8(%rax)

> <rcx=J                7fffef45c000
> <rax=J                7fffef45c0f8

> <rcx::whatis
7fffef45c000 is in /usr/lib/brand/lx/amd64/lx_vdso.so.1 [7fffef45c000,7fffef45f000)
> <rax::whatis
7fffef45c0f8 is in /usr/lib/brand/lx/amd64/lx_vdso.so.1 [7fffef45c000,7fffef45f000)
```

Generate a core (in this case by running `id` inside the zone) and load it into gdb:
```
bloody# coreadm -G default+debug

lx# /usr/bin/id
Memory fault(coredump)

bloody# gdb /zones/lx/root/usr/lib64/ld-linux-x86-64.so.2 --core 0-id-18465
(gdb) info address dl_main
Symbol "dl_main" is a function at address 0x21e30.
^Z
bloody# dc
16 d i o 21E30 24C7 + p
242F7
bloody# fg
(gdb) info line *0x242F7
Line 83 of "/usr/src/debug/glibc-2.34-29.el9.x86_64/elf/get-dynamic-info.h"
   starts at address 0x242f7 <dl_main+9415> and ends at 0x242fb <dl_main+9419>.
```

which corresponds to this C code:

```C
  /* Don't adjust .dynamic unnecessarily.  */
  if (l->l_addr != 0 && dl_relocate_ld (l))
    {
      ElfW(Addr) l_addr = l->l_addr;

# define ADJUST_DYN_INFO(tag) \
      do                                                                      \
        if (info[tag] != NULL)                                                \
         info[tag]->d_un.d_ptr += l_addr;                                     \
      while (0)

      ADJUST_DYN_INFO (DT_HASH);
      ADJUST_DYN_INFO (DT_PLTGOT);
      ADJUST_DYN_INFO (DT_STRTAB);
      ADJUST_DYN_INFO (DT_SYMTAB);
```

```asm
ld-linux-x86-64.so.2`dl_main+0x24b0:    testb  $0x20,0x31e(%r13)
ld-linux-x86-64.so.2`dl_main+0x24b8:    jne    -0x117f  <ld-linux-x86-64.so.2`dl_main+0x133f>
ld-linux-x86-64.so.2`dl_main+0x24be:    movq   0x60(%r13),%rax
ld-linux-x86-64.so.2`dl_main+0x24c2:    testq  %rax,%rax
ld-linux-x86-64.so.2`dl_main+0x24c5:    je     +0x4     <ld-linux-x86-64.so.2`dl_main+0x24cb>
ld-linux-x86-64.so.2`dl_main+0x24c7:    addq   %rcx,0x8(%rax)                                <---- crash here
ld-linux-x86-64.so.2`dl_main+0x24cb:    movq   0x58(%r13),%rax
ld-linux-x86-64.so.2`dl_main+0x24cf:    testq  %rax,%rax
ld-linux-x86-64.so.2`dl_main+0x24d2:    je     +0x4     <ld-linux-x86-64.so.2`dl_main+0x24d8>
ld-linux-x86-64.so.2`dl_main+0x24d4:    addq   %rcx,0x8(%rax)
ld-linux-x86-64.so.2`dl_main+0x24d8:    movq   0x68(%r13),%rax
ld-linux-x86-64.so.2`dl_main+0x24dc:    testq  %rax,%rax
ld-linux-x86-64.so.2`dl_main+0x24df:    je     +0x4     <ld-linux-x86-64.so.2`dl_main+0x24e5>
```

```
lx# truss /usr/bin/id
...
brand(147, 0x7FFFEEF6F000, 0x00000002, 0x7FFFEEF703F0, 0x00000000) = 0
getcontext(0x7FFFFFEFF830)
brand(139, 0x7FFFFFEFF830, 0x00000000, 0x00000000, 0x000003FF) = 0
    Incurred fault #6, FLTBOUNDS  %pc = 0x7FFFEF2E92F7
      siginfo: SIGSEGV SEGV_ACCERR addr=0x7FFFEF45C100
    Received signal #11, SIGSEGV [default]
      siginfo: SIGSEGV SEGV_ACCERR addr=0x7FFFEF45C100
```

```
> ::mappings
            BASE            LIMIT             SIZE NAME
    7fffeeef0000     7fffeeef1000             1000 [ anon ]
    7fffeeef3000     7fffeeef4000             1000 [ anon ]
    7fffeef00000     7fffeef06000             6000 [ anon ]
    7fffeef10000     7fffeef11000             1000 [ anon ]
    7fffeef20000     7fffeefab000            8b000 /lib/amd64/libnsl.so.1
    7fffeefbb000     7fffeefbe000             3000 /lib/amd64/libnsl.so.1
    7fffeefbe000     7fffeefc6000             8000 /lib/amd64/libnsl.so.1
    7fffeefd0000     7fffeefd1000             1000 [ anon ]
    7fffeefe0000     7fffeefe9000             9000 /lib/amd64/librpcsvc.so.1
    7fffeeff9000     7fffeeffa000             1000 /lib/amd64/librpcsvc.so.1
    7fffef000000     7fffef001000             1000 [ anon ]
    7fffef010000     7fffef197000           187000 /lib/amd64/libc.so.1
    7fffef1a7000     7fffef1b3000             c000 /lib/amd64/libc.so.1
    7fffef1b3000     7fffef1b7000             4000 /lib/amd64/libc.so.1
    7fffef1c0000     7fffef1c1000             1000 [ anon ]
    7fffef1d0000     7fffef1d2000             2000 /usr/lib/amd64/libmapmalloc.so.1
    7fffef1e2000     7fffef1e3000             1000 /usr/lib/amd64/libmapmalloc.so.1
    7fffef1f0000     7fffef1f1000             1000 /usr/lib/brand/lx/ld/64/ld.config
    7fffef200000     7fffef201000             1000 [ anon ]
    7fffef210000     7fffef211000             1000 [ anon ]
    7fffef220000     7fffef221000             1000 [ anon ]
    7fffef230000     7fffef231000             1000 [ anon ]
    7fffef240000     7fffef241000             1000 [ anon ]
    7fffef24b000     7fffef24d000             2000 /zones/lx/root/usr/lib64/ld-linux-x86-64.so.2
    7fffef24d000     7fffef273000            26000 /zones/lx/root/usr/lib64/ld-linux-x86-64.so.2
    7fffef273000     7fffef27e000             b000 /zones/lx/root/usr/lib64/ld-linux-x86-64.so.2
    7fffef27f000     7fffef283000             4000 /zones/lx/root/usr/lib64/ld-linux-x86-64.so.2
    7fffef284000     7fffef2bc000            38000 /zones/lx/root/usr/lib/systemd/systemd
    7fffef2bc000     7fffef3ab000            ef000 /zones/lx/root/usr/lib/systemd/systemd
    7fffef3ab000     7fffef411000            66000 /zones/lx/root/usr/lib/systemd/systemd
    7fffef411000     7fffef45b000            4a000 /zones/lx/root/usr/lib/systemd/systemd
    7fffef45c000     7fffef45f000             3000 /usr/lib/brand/lx/amd64/lx_vdso.so.1
    7fffef460000     7fffef461000             1000 [ unknown ]                    <-- comm page?
    7fffef462000     7fffef4b6000            54000 /lib/amd64/ld.so.1
    7fffef4c6000     7fffef4c9000             3000 /lib/amd64/ld.so.1
    7fffef4c9000     7fffef4cb000             2000 /lib/amd64/ld.so.1
    7fffef4cc000     7fffef4ea000            1e000 /usr/lib/amd64/lx_brand.so.1
    7fffef4fa000     7fffef4fe000             4000 /usr/lib/amd64/lx_brand.so.1
    7fffef4fe000     7fffef4ff000             1000 /usr/lib/amd64/lx_brand.so.1
    7fffffefd000     7ffffff00000             3000 [ stack ]
```

### PR DESCRIPTION
**Pull request by citrus-it**
This fixes #1142

### CODE REVIEW

**Comment by danmcd** [comment]
Make sure you confirm after a nightly that the lx_vdso binaries have the right elfdump output.  Once I look at the code I'll patch illumos-joyent with this as well.

**Review by danmcd — APPROVED** [review]
I'll pull this for an illumos-joyent/smartos-live  build.

**Review by alslater — COMMENTED** [review]
PY_DYNAMIC typo on line 359?

**Review by oetiker — DISMISSED** [review]

**Review by danmcd — APPROVED** [review]
One nit which you can dismiss if you wish.  Approving because it's merely an editorial nit, and the technical content is still good.

File: usr/src/lib/brand/lx/lx_vdso/tools/vdso_tool.c [hunk 1]
```diff
@@ -11,16 +11,18 @@
...
 /*
  * vdso_tool: a build-time tool for adjusting properties of the "lx_vdso.so.1"
  * object we build for VDSO emulation in the LX brand.
  *
- * This tool ensures that the shared object contains only one loadable program
- * header (PT_LOAD), and extends the size of that program header to induce the
- * loading of all sections into memory.  It also sets a few attributes in the
- * ELF header.
+ * This tool sets a few attributes in the ELF header, ensures that the
+ * PT_DYNAMIC section's p_flags do not have the writeable bit set, ensures that
```
[the remark this answers is not in the document]
Comment (citrus-it) [reply usr/src/lib/brand/lx/lx_vdso/tools/vdso_tool.c]: Thanks - I've just restructured the whole comment to bullet points to make it all a bit more readable and complete.

**Review by hadfl — APPROVED** [review]

**Pull request merged**